Initial strict runtime check

naucke opened the pull request:
no datatypes/lambdas yet

naucke committed:
Initial strict runtime check

naucke commented on src/Agda2Hs/Compile/Data.hs:
```diff
@@ -55,17 +61,30 @@ compileData newtyp ds def = do
+
+  let qnameList = qnameToList c
+      -- not the datatype itself
+      qualifier = NE.take (length qnameList - 2) qnameList
```
This whole `let` block exists to generate appropriate smart constructor names and I find it a bit ugly, especially the marked line which exists to export the qualified name of the smart constructor _without_ including the name of the data type in the constructor. Is there a better way to do this?

naucke commented on src/Agda2Hs/Compile/RuntimeCheckUtils.hs:
```diff
@@ -0,0 +1,108 @@
+-- Takes telescope of dependent type to check, name, and expression on success.
+-- TODO name conflicts
+-- TODO togglable
+checkRtc :: Tele (Dom Type) -> QName -> Hs.Exp () -> C (Maybe (Hs.Decl ()))
```
This calls all main functionality. It doesn't handle nested erased types yet.

naucke commented on src/Agda2Hs/Compile/RuntimeCheckUtils.hs:
```diff
@@ -0,0 +1,108 @@
+        )
+        chks
+    successApp = eApp success $ map hsVar args
+    rhs = Hs.GuardedRhss () (Hs.GuardedRhs () [Hs.Qualifier () andChks] successApp : errChks)
```
This is all a bit wordy with the Haskell constructs but I suppose there isn't really a way around it.

naucke commented on src/Agda2Hs/Compile/ClassInstance.hs:
```diff
@@ -321,15 +317,6 @@ findDefinitions p m = do
   map snd <$> filterM (uncurry p) inMod
 
 
-resolveStringName :: String -> C QName
```
Moved to utils.

naucke commented on src/Agda2Hs/Compile/RuntimeCheckUtils.hs:
```diff
@@ -0,0 +1,108 @@
+  dec <- decify t
+  sth <- findDecInstances dec
+  -- XXX functor something?
+  findDecInstances dec >>= \case
+    Just r -> Just <$> compileTerm dec r
+    Nothing -> return Nothing
```
This really feels to me like an fmap should be possible, but I don't know how.

liesnikov replied:
something like `(fmap $ compileTerm dec r) =<< findDecInstances dec`?

liesnikov replied:
`copileTerm dec <$> sequence (findDecInstanes dec)`, I think?

naucke replied:
Nope, `sequence (findDecInstances dec)` expects a `C (C Term)` (idk why exactly), but is a `Maybe (C Term)`

liesnikov replied:
"expects" as in "has the type"?

naucke replied:
Just with `sth <- sequence $ findDecInstances dec` (to avoid confusion with the extra error messages that appear with your suggested fmap):
```
    • Couldn't match type ‘Maybe’ with ‘C’
      Expected: C (C Term)
        Actual: Maybe (C Term)
    • In a stmt of a 'do' block: sth <- sequence $ findDecInstances dec
```
(where I replaced
```
                  Control.Monad.Trans.RWS.CPS.RWST
                  Agda2Hs.Compile.Types.CompileEnv
                  Agda2Hs.Compile.Types.CompileOutput
                  Agda2Hs.Compile.Types.CompileState
                  Agda.TypeChecking.Monad.Base.TCM
```
with `C` for brevity)

naucke replied:
But how can I not be in a `C` block when I need to call `compileTerm :: Type -> Term -> C (Exp ())`?

liesnikov replied:
the result of `copileTerm dec <$> sequence (findDecInstanes dec)` is a `Maybe (C whatevs)`, if you want to bind it in a `C` do-block I guess you can do another `sequence` on the result?

Alternatively you could try doing
```
r1 <- findDecInstances dec
traverse (compileTerm dec) r1
```

Or maybe `mapMM` ([hackage](https://hackage.haskell.org/package/Agda-2.6.4.3/docs/Agda-Utils-Monad.html#v:mapMM))is what you're after here, to get something like `mapMM (compileTerm dec) (findDecInstances dec)`

naucke replied:
`traverse`, I was looking for `traverse`. Can even just `findDecInstances dec >>= traverse (compileTerm dec)`. Thank you! Update coming.

liesnikov replied:
https://impurepics.com/posts/2020-10-03-always-traverse.html

naucke commented on src/Agda2Hs/Compile/Name.hs:
```diff
@@ -102,7 +102,8 @@ compileQName f
     let mod0 = qnameModule $ fromMaybe f parent
-    mod <- compileModuleName mod0
+    mod <- compileModuleName mod0 >>= \case
+      Hs.ModuleName () s -> return $ Hs.ModuleName () $ s ++ ".PostRtc"
```
This feels a bit error-prone. Is there a better way to do this?

liesnikov commented:
No, the error is because you're using it in a C-do block. The compiler is actually suggesting that the term has the right type: `Actual: Maybe (C Term)`
-------- Original message --------From: Jakob Naucke ***@***.***> Date: 19/04/2024  19:19  (GMT+01:00) To: jakobn-ai/agda2hs ***@***.***> Cc: Bohdan Liesnikov ***@***.***>, Comment ***@***.***> Subject: Re: [jakobn-ai/agda2hs] Initial strict runtime check (PR #2) 
@jakobn-ai commented on this pull request.



In src/Agda2Hs/Compile/RuntimeCheckUtils.hs:
> +  findDecInstances dec >>= \case
+    Just r -> Just <$> compileTerm dec r
+    Nothing -> return Nothing

Just with sth <- sequence $ findDecInstances dec (to avoid confusion with the extra error messages that appear with your suggested fmap):
    • Couldn't match type ‘Maybe’ with ‘C’
      Expected: C (C Term)
      Actual: Maybe (C Term)
    • In a stmt of a 'do' block: sth <- sequence $ findDecInstances dec

(where I replaced
                  Control.Monad.Trans.RWS.CPS.RWST
                  Agda2Hs.Compile.Types.CompileEnv
                  Agda2Hs.Compile.Types.CompileOutput
                  Agda2Hs.Compile.Types.CompileState
                  Agda.TypeChecking.Monad.Base.TCM

with C for brevity)

—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you commented.Message ID: ***@***.***>

naucke committed:
Add OverloadedStrings to Record for simplicity

naucke committed:
Move resolveStringName to utils and make failable

for greater reuse

jespercockx commented on src/Agda2Hs/Compile.hs:
```diff
@@ -60,20 +62,22 @@ moduleSetup _ _ m _ = do
   :: Options -> ModuleEnv -> IsMain -> Definition 
-  -> TCM (CompiledDef, CompileOutput)
-compile opts tlm _ def =
+  -> TCM ((CompiledDef, CompiledDef), CompileOutput)
```
Could you add a comment to explain what the two definitions are?

jespercockx replied:
Perhaps it would be worth it to define a dedicated record type for pairing the two versions of a definition.

jespercockx commented on src/Agda2Hs/Compile/Data.hs:
```diff
@@ -37,14 +44,23 @@ compileData newtyp ds def = do
+      -- Always export data type name
+      tellNoErased $ prettyName ++ "(" ++ intercalate ", " noneErased ++ ")"
+      return chks)
+      $ return []
```
use `ifNotM` to avoid trailing `return`s like this.

jespercockx commented on src/Agda2Hs/Compile/Data.hs:
```diff
@@ -37,14 +44,23 @@ compileData newtyp ds def = do
+    chkdCs <- mapM (compileConstructor params) cs
+
+    chks <- ifM (emitsRtc $ defName def) (do
+      let (noneErased, chks) = mapSnd concat $ partitionEithers $ mapMaybe snd chkdCs
```
Unclear to me what exactly this code is doing, are you somehow adding the precondition checks from the datatype to the constructors? I don't think that's a useful/necessary thing to do.

naucke replied:
Each constructor receives a smart constructor according to its (the constructor's) type, but maybe I don't understand you right here?

Inside the `do` block, the names to be exported are written (always data type plus the constructors with no erased types); all checkable constructors are returned as `chks`.

jespercockx commented on src/Agda2Hs/Compile/Data.hs:
```diff
@@ -55,17 +71,30 @@ compileData newtyp ds def = do
 
-compileConstructor :: [Arg Term] -> QName -> C (Hs.QualConDecl ())
+compileConstructor :: [Arg Term] -> QName -> C (Hs.QualConDecl (),
+{- optional exportable cons/checker function -} Maybe (Either String [Hs.Decl ()]))
```
Try to avoid nested usages of `Maybe` and `Either` like this, usually it's a good sign that you should define a data type yourself.

jespercockx commented on src/Agda2Hs/Compile/Function.hs:
```diff
@@ -162,7 +163,16 @@ compileFun' withSig def@Defn{..} = inTopContext $ withCurrentModule m $ do
+        typeTel <- theTel <$> telView defType
+        let success = hsVar $ prettyShow m ++ ".PostRtc." ++ prettyShow n
+        checkRtc typeTel defName success >>= \case
+          NoneErased -> do tellNoErased $ prettyShow n; return []
```
This pattern match on the result of `checkRtc` comes back a few times with a very similar structure, perhaps it's worth encapsulating in a helper function.

naucke replied:
Hmm, there are subtle differences though which is why I designed the signatures in this way:

|           | None erased                 | All checkable               |
|-----------|-----------------------------|-----------------------------|
| Function  | mark for export             | prepend signature (list)    |
| Data type | mark constructor for export | prepend signature (element) |
| Record    | mark constructor for export | have no signature           |

jespercockx commented on src/Agda2Hs/Compile/Function.hs:
```diff
@@ -376,13 +386,19 @@ withClauseLocals curModule c@Clause{..} k = do
     [Hs.TypeDecl _ hd b] ->
       checkTransparentTypeDef hd b
-    _ -> __IMPOSSIBLE__
+    _ -> __IMPOSSIBLE__) . fst
```
Move `fst` to the `compileFun` call above like this: `(fst <$> compileFun False def) >>= ...` (parenthesis around case expressions are ugly)

jakobn-ai replied:
`hlint` prefers this, which I agree doesn't make a lot of sense in this instance. I have moved to a separate assignment (also in the case of `RuntimeCheckUtils.checkRtc`)

naucke replied:
`hlint` prefers this, which I agree doesn't make much sense here. I have moved to a separate assignment (also for `RuntimeCheckUtils.checkRtc`)
(sent single comment earlier by accident, sorry for the noise)

jespercockx commented on src/Agda2Hs/Compile/Record.hs:
```diff
@@ -49,7 +54,7 @@ compileMinRecord fieldNames m = do
   -- * it's using the fields and definitions from the minimal record and not the parent record
   compiled <- withMinRecord m $ addContext (defaultDom rtype) $ compileLocal $
-    fmap concat $ traverse (compileFun False) defaults
+    concat <$> traverse (fmap (uncurry (++)) . compileFun False) defaults
```
Why is it fine to just concatenate the pre- and post-rtc functions here?

naucke replied:
This is only called for classes, which are prohibited. However, it indeed makes more sense to have `fst` (or `defn` in the update with a data type).

jespercockx commented on src/Agda2Hs/Compile/Record.hs:
```diff
@@ -100,24 +105,41 @@ compileRecord target def = do
   where
-    rName = hsName $ prettyShow $ qnameName $ defName def
-    cName | recNamedCon = hsName $ prettyShow $ qnameName $ conName recConHead
-          | otherwise   = rName   -- Reuse record name for constructor if no given name
```
Please preserve the comment

jespercockx commented on src/Agda2Hs/Compile/RuntimeCheckUtils.hs:
```diff
@@ -0,0 +1,252 @@
+telify = theTel . telView' . snd . unDom
+
+-- Check a type is erased and at top level; in this case, it should be checked.
+checkTopLevelErased :: Dom (a, Type) -> Bool
```
This is somewhat problematic: since https://github.com/agda/agda2hs/pull/304/ `agda2hs` no longer purely relies on erasure annotations to determine whether to drop an argument.
Ideally, for consistency with the rest of agda2hs we would use the result of `compileDom` to determine whether or not an argument is erased. Otherwise, there is still a risk we miss preconditions in `Prop`, for example

jespercockx commented on src/Agda2Hs/Compile/RuntimeCheckUtils.hs:
```diff
@@ -0,0 +1,252 @@
+            chkName = hsName $ prettyShow $ qnameName name
+            chk = createRtc chkName theirLhs theirChks rhs $ binds theirDecls
+        return $ Checkable [chk]
+      . snd
```
Weird layout, I'm not even sure *what* you are composing the `snd` with

naucke replied:
`hlint` prefers this, which I agree doesn't make much sense here. I have moved to a separate assignment (also for `Function.checkTransparentPragma`)

jespercockx commented on src/Agda2Hs/Compile/RuntimeCheckUtils.hs:
```diff
@@ -0,0 +1,252 @@
+
+-- Gather telescope from domain
+telify :: Dom (a, Type) -> Telescope
+telify = theTel . telView' . snd . unDom
```
Why is it okay to use `telView'` here instead of `telView`?

naucke replied:
(as discussed in person) it should be fine to always have `telView'` because we don't need type reduction. I have updated the rest of the PR accordingly.

jespercockx commented on src/Agda2Hs/Compile/RuntimeCheckUtils.hs:
```diff
@@ -0,0 +1,252 @@
+-- Failably find instances for decidable terms
+findDecInstances :: Type -> C (Maybe Term)
+findDecInstances t =
+  liftTCM $
```
Typically I would put both `liftTCM` and `do` on the same line as `findDecInstances t =`, saving on indentation and vertical space.

naucke replied:
Hmmm, I really enjoyed working with autoformatters and all the major ones seem to do this. I've moved the `liftTCM` to the caller, but let me know if you find that unclear.

jespercockx commented on src/Agda2Hs/Compile/RuntimeCheckUtils.hs:
```diff
@@ -0,0 +1,252 @@
+    -- Partition out arguments that are erased and at top level (those we will attempt to check)
+    (topLevelErased, call) = partition (checkTopLevelErased . fst) $ zip doms telsUpTo
+
+-- Check a single type for runtime checkability.
```
Since this is the core logic of this PR, I would appreciate some more detailed commentary with possibly an example to make clear what the `NameIndices` are for. Adding line-by-line explanations to the function definitions would also be a great help, since they are quite difficult to grasp at the moment.

naucke replied:
I have enhanced the comments, do let me know if you feel this is sufficiently illustrated.

jespercockx reviewed (commented):
Please find some of my comments on this PR below. I think it's a great addition, however it also adds quite a bit extra complexity to different parts of the `agda2hs` code base. So it would be great to have the opinion of some others (maybe @omelkonian and @ulfnorell) to hear whether they think this is something worth adding (and continue supporting in the future).

jespercockx committed:
[ fix #119 ] Throw an error when an unknown name appears in generated Haskell code

jespercockx committed:
[ re #354 ] Do not check canonicity of erased instances

jespercockx committed:
Lower version bound of aeson to 2.0.3

jespercockx committed:
Add support for GHC 9.10.2

jespercockx committed:
Update GHC and Cabal versions for CI

jespercockx committed:
[ re #357 ] Generalize + simplify rule for compiling lambdas to partial type applications

jespercockx committed:
[ re #357 ] Erase type arguments of `mapDec`

The types a and b do not appear in Haskell code, so they should be marked as erased

jespercockx committed:
Add change log for agda2hs 1.3

jespercockx committed:
Add new contributors to the license file

jespercockx committed:
Remove (unmaintained) Smuggler tool from cabal file

jespercockx committed:
Add missing upper bound for mtl package

jespercockx committed:
Update author list in agda2hs.cabal

liesnikov committed:
remove the jailbreak now that aeson bounds are relaxed

jespercockx committed:
[ #308 ] Already fixed by new canonicity check, adding test case

flupe committed:
[ fix #370 ] add locate command, data-files in cabal config

flupe committed:
update installation instructions

flupe committed:
update changelog

anuyts committed:
Docs: tweak features.md
